### ISSUE TITLE
Describe all types of deltas

### ISSUE THREAD

**Issue by dslmeinte** [LionWeb-io/specification#280]
Get inspired by the list of messages in issue #28 , and put a description of all types of deltas in _this_ issue.

# Out of scope (for now)
* language migration
* versioning
* live collaboration (e.g. "user A currently edits this node")

# Open questions
* Do we send (all) useful info with event or only absolutely needed? (Example: OldIndex) #287 
* Do we send a whole node or only its id?
* If we send `removeChild()`, do we also need to send that for all descendants? #286 

# Events
## Parameter types
* Node: Either the whole node as json, or its id (tbd)
* TargetNode: Target's node as id
* MetaPointer: MetaPointer-tuple to describe an M2 element (classifier, feature, etc.)
* PropertyValue: Property value encoded as string
* ResolveInfo: ResolveInfo of a reference as string
* Index: Position inside a multiple link or annotations as integer

## Partitions
| Name | Required Params | Optional Params |
|--------|--------|--------|
| addPartition | newPartition: _Node_ |  |
| removePartition | removedPartition: _Node_ or _TargetNode_ |  |

## Nodes
| Name | Required Params | Optional Params |
|--------|--------|--------|
| changeConcept | node: _TargetNode_<br>newConcept: _MetaPointer_ | oldConcept: _MetaPointer_ |

## Properties
| Name | Required Params | Optional Params |
|--------|--------|--------|
| addProperty | node: _TargetNode_<br>property: _MetaPointer_<br>newValue: _PropertyValue_ |  |
| removeProperty | node: _TargetNode_<br>property: _MetaPointer_ | oldValue: _PropertyValue_ |
| changeProperty | node: _TargetNode_<br>property: _MetaPointer_<br>newValue: _PropertyValue_ | oldValue: _PropertyValue_  |

## Children
| Name | Required Params | Optional Params |
|--------|--------|--------|
| addChild | parent: _TargetNode_<br>newChild: _Node_<br>containment: _MetaPointer_<br>index: _Index_ |  |
| removeChild | removedChild: _Node_ or _TargetNode_| parent: _TargetNode_<br>containment: _MetaPointer_<br>index: _Index_  |
| replaceChild | // implies `newChild` must be new<br>newChild: _Node_<br>replacedChild: _Node_ or _TargetNode_ | parent: _TargetNode_ <br>containment: _MetaPointer_<br>index: _Index_ |
| moveChildFromOtherContainment | newParent: _TargetNode_<br>newContainment: _MetaPointer_<br>newIndex: _Index_<br>movedChild: _TargetNode_ | oldParent: _TargetNode_ <br>oldContainment: _MetaPointer_<br>oldIndex: _Index_ |
| moveChildFromOtherContainmentInSameParent| newContainment: _MetaPointer_<br>newIndex: _Index_<br>movedChild: _TargetNode_ | parent: _TargetNode_ <br>oldContainment: _MetaPointer_<br>oldIndex: _Index_ |
| moveChildInSameContainment| newIndex: _Index_<br>movedChild: _TargetNode_ | parent: _TargetNode_ <br>containment: _MetaPointer_<br>oldIndex: _Index_ |

## Annotations
| Name | Required Params | Optional Params |
|--------|--------|--------|
| addNewAnnotation| parent: _TargetNode_<br>newAnnotation: _Node_<br>index: _Index_ |  |
| removeAnnotation | removedAnnotation: _Node_ or _TargetNode_| parent: _TargetNode_<br>index: _Index_  |
| replaceAnnotation | // implies `newAnnotation` must be new<br>newAnnotation: _Node_<br>replacedAnnotation: _Node_ or _TargetNode_ | parent: _TargetNode_ <br>index: _Index_ |
| moveAnnotationFromOtherParent | newParent: _TargetNode_<br>newIndex: _Index_<br>movedAnnotation: _TargetNode_ | oldParent: _TargetNode_ <br>oldIndex: _Index_ |
| moveAnnotationInSameParent| <br>newIndex: _Index_<br>movedAnnotation: _TargetNode_ | parent: _TargetNode_ <br>oldIndex: _Index_ |

## References
| Name | Required Params | Optional Params |
|--------|--------|--------|
| addReference | parent: _TargetNode_<br>newTarget: _TargetNode_<br>newResolveInfo: _ResolveInfo_<br>reference: _MetaPointer_<br>index: _Index_ |  |
| removeReference | parent: _TargetNode_<br>reference: _MetaPointer_<br>index: _Index_ | removedTarget: _TargetNode_<br>removedResolveInfo: _ResolveInfo_ |
| changeReference | parent: _TargetNode_<br>newTarget: _TargetNode_<br>newResolveInfo: _ResolveInfo_<br>reference: _MetaPointer_<br>index: _Index_ | replacedTarget: _TargetNode_<br>replacedResolveInfo: _ResolveInfo_ |
| addReferenceResolveInfo | parent: _TargetNode_<br>newResolveInfo: _ResolveInfo_<br>reference: _MetaPointer_<br>index: _Index_ |  |
| removeReferenceResolveInfo | parent: _TargetNode_<br>reference: _MetaPointer_<br>index: _Index_ | removedResolveInfo: _ResolveInfo_ |
| changeReferenceResolveInfo | parent: _TargetNode_<br>newResolveInfo: _ResolveInfo_<br>reference: _MetaPointer_<br>index: _Index_ | target: _TargetNode_<br>replacedResolveInfo: _ResolveInfo_ |
| addReferenceTargetId | parent: _TargetNode_<br>newTargetId: _TargetNode_<br>reference: _MetaPointer_<br>index: _Index_ |  |
| removeReferenceTargetId | parent: _TargetNode_<br>reference: _MetaPointer_<br>index: _Index_ | removedTargetId: _TargetNode_ |
| changeReferenceTargetId | parent: _TargetNode_<br>newTargetId: _TargetNode_<br>reference: _MetaPointer_<br>index: _Index_ | resolveInfo: _ResolveInfo_<br>replacedTargetId: _TargetNode_ |

## Composite
| Name | Required Params | Optional Params |
|--------|--------|--------|
| composite | parts: _Event[]_ |  |
| giveUpDeltasJustReload | |  |

## High-level
* undo
* redo

**Comment by enikao** [comment]
Depends on #273 and #284

**Comment by enikao** [comment]
Writing this table in Markdown became too tedious, so I started with an AsciiDoctor version: https://github.com/LionWeb-io/specification/pull/288 / https://github.com/LionWeb-io/specification/blob/3d05a14197189b809a17db9d889f78be1f8d2b79/delta/delta-api.adoc

**Comment by enikao** [comment]
Will add all 3 variants of `moveReference`
